batch2/batch/driver.py

```diff
@@ -0,0 +1,470 @@
+                async with f(session) as resp:
+                    if self.instance:
+                        self.instance.mark_as_healthy()
+                    return resp, None
+        except asyncio.CancelledError:  # pylint: disable=try-except-raise
+            raise
+        except Exception as err:  # pylint: disable=broad-except
+            if self.instance:
+                self.instance.mark_as_unhealthy()
+            return None, err
+
+    async def _post(self, url, json=None):
+        return await self._request(lambda session: session.post(url, json=json))
+
+    async def _get(self, url):
+        return await self._request(lambda session: session.get(url))
+
+    async def create(self):
+        async with self.lock:
+            assert not self.on_ready
+
+            config = await self.config()
+
+            if self.deleted:
+                log.info(f'pod already deleted {self.name}')
+                return
+
+            if not self.instance:
+                log.info(f'instance was deactivated before {self.name} could be created; rescheduling')
+                asyncio.ensure_future(self.put_on_ready())
+                return
+
+            inst = self.instance
+            url = f'http://{inst.ip_address}:5000/api/v1alpha/pods/create'
+            resp, err = await self._post(url, config)
+            if resp:
+                if resp.status == 200:
+                    log.info(f'created {self.name} on inst {inst}')
+                    return None
+                else:
+                    log.info(f'failed to create {self.name} on inst {inst} due to {resp}')
+                    return
+            else:
+                assert err
+                log.info(f'failed to create {self.name} on inst {inst} due to {err}, putting back on ready queue')
+                asyncio.ensure_future(self.put_on_ready())
+
+    async def delete(self):
+        assert self.deleted
+
+        async with self.lock:
+            inst = self.instance
+            if inst:
+                url = f'http://{inst.ip_address}:5000/api/v1alpha/pods/{self.name}/delete'
+                resp, err = await self._post(url)
+                if resp:
+                    if resp.status == 200:
+                        log.info(f'deleted {self.name} from inst {inst}')
+                    else:
+                        log.info(f'failed to delete {self.name} from inst {inst} due to {resp}')
+                else:
+                    assert err
+                    log.info(f'failed to delete {self.name} on inst {inst} due to err {err}, ignoring')
+
+            await self.unschedule()
+            asyncio.ensure_future(db.pods.delete_record(self.name))
+
+    async def read_pod_log(self, container):
+        assert container in tasks
+        log.info(f'reading container log for {self.name}, {container} from instance {self.instance}')
+
+        if self.instance is None:
+            return None, None
+
+        inst = self.instance
```

Comment:
This pattern should get wrapped up in a function. Longer term we should probably use a tiny class to wrap up the logic of making HTTP requests to a worker instance.